pkg/loop/telem.go

```diff
@@ -2,34 +2,118 @@ package loop
-func SetupTelemetry(registerer prometheus.Registerer) GRPCOpts {
-	otel.SetTracerProvider(sdktrace.NewTracerProvider(
-		sdktrace.WithSampler(sdktrace.AlwaysSample()),
-	))
-	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+type TracingConfig struct {
+	// NodeAttributes are the attributes to attach to traces.
+	NodeAttributes map[string]string
+
+	// Enables tracing; requires a collector to be provided
+	Enabled bool
 
+	// Collector is the address of the OTEL collector to send traces to.
+	CollectorTarget string
+
+	// SamplingRatio is the ratio of traces to sample. 1.0 means sample all traces.
+	SamplingRatio float64
+}
+
+// NewGRPCOpts initializes open telemetry and returns GRPCOpts with telemetry interceptors.
+// It is called from the host and each plugin - intended as there is bidirectional communication
+func NewGRPCOpts(registerer prometheus.Registerer) GRPCOpts {
 	if registerer == nil {
 		registerer = prometheus.DefaultRegisterer
 	}
 	return GRPCOpts{DialOpts: dialOptions(registerer), NewServer: newServerFn(registerer)}
 }
 
+// SetupTracing initializes open telemetry with the provided config.
+// It sets the global trace provider and opens a connection to the configured collector.
+func SetupTracing(config TracingConfig) error {
+	if !config.Enabled {
+		return nil
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+
+	// Enough to shutdown the underlying connection since DialContext is used in blocking mode
+	defer cancel()
+	conn, err := grpc.DialContext(ctx, config.CollectorTarget,
+		// Note the use of insecure transport here. TLS is recommended in production.
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		// safe usage in this case https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md#dialing-in-grpc
+		grpc.WithBlock(),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Set up a trace exporter
+	// Shutting down the traceExporter will not shutdown the underlying connection.
+	traceExporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
+	if err != nil {
+		return err
+	}
+
+	var version string
+	var service string
+	buildInfo, ok := debug.ReadBuildInfo()
+	if !ok {
+		version = "unknown"
+		service = "cl-node"
+	} else {
+		version = buildInfo.Main.Version
+		service = buildInfo.Main.Path
+	}
+
+	attributes := []attribute.KeyValue{
+		semconv.ServiceNameKey.String(service),
+		semconv.ProcessPIDKey.Int(os.Getpid()),
+		semconv.ServiceVersionKey.String(version),
+	}
+
+	for k, v := range config.NodeAttributes {
+		attributes = append(attributes, attribute.String(k, v))
+	}
+
+	resource := resource.NewWithAttributes(
+		semconv.SchemaURL,
+		attributes...,
+	)
+
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithBatcher(traceExporter),
+		sdktrace.WithResource(resource),
+		sdktrace.WithSampler(sdktrace.TraceIDRatioBased(config.SamplingRatio)),
```

Comment:
Should we consider `ParentBased`? https://pkg.go.dev/go.opentelemetry.io/otel/sdk@v1.19.0/trace#ParentBased

Reply:
Absolutely, though for right now since there is no active span when the GRPC interceptors are invoked (on the client side of the request), this will have no effect if I pass the `TraceIDRatioBased` sampler in as the root sampler. Or were you imagining something different?

Reply:
Whatever side initiates the request.

Reply:
How can it be the case that we are only tracing on one side? What happens when the server side calls back in to a client with the same context?

Reply:
My point is just that as the code is written right now, there is no parent span being passed into the interceptors. So, when a request is initiated, the client is generating a fresh span and trace that is extracted on the server side. The span and trace being generated means nothing if the server side does not decide to sample that trace. In the code as is, TraceIDRatioBased samples a percentage of traces based on the trace ID. We currently exert 0 control over the trace IDs, so this is effectively random. 

`ParentBased`sampling will honor the sampling decision for the parent span. Including this today will not affect sampling on the server side because there is no parent span.

Reply:
> The span and trace being generated means nothing if the server side does not decide to sample that trace.

Isn't the point of using `ParentBased` to control that decision? https://github.com/open-telemetry/opentelemetry-go/blob/sdk/v1.19.0/sdk/trace/sampling.go#L209

Reply:
> This is the most popular sampler and is the one recommended by the official OpenTelemetry [documentation](https://opentelemetry.io/docs/go/exporting_data/#sampling). 
>
> When a trace begins we make a decision whether or not to sample it. Whatever the decision, it is bubbled down inside the process and in other processes(for example in other services).
>
> There is a notable advantage to this: you always get the complete picture.

https://www.aspecto.io/blog/opentelemetry-sampling-everything-you-need-to-know/#parent-based-sampler

Reply:
There are 4 options here based on whether the local or remote parent is or is not sampled (`[WithLocalParentNotSampled](https://github.com/open-telemetry/opentelemetry-go/blob/sdk/v1.19.0/sdk/trace/sampling.go#L253)` for example) - my understanding is that the current code does not have a parent, so these options will have no impact.

Reply:
When you say 'current code', is that making assumptions outside of this package?

Reply:
Yes - we're currently not doing anything to chain spans across plugin invocations. Once that changes and we track some series of plugins calls (like in a Job), then I believe the Parent based rules will become useful

Reply:
> Yes

We shouldn't. This is a general library at the bottom of the dependency tree. It does not know anything about how it is currently used upstream or will be used in the future. We should aim to be maximally flexible by default. Why leave this out if we need to circle back later?

> we're currently not doing anything to chain spans across plugin invocations.

What would we need to do, besides pass a context?

Reply:
> Why leave this out if we will need to circle back later?

It feels like we're going in circles here, but it's probably my bad. I have no problem with including or not including this. It makes no difference to me. We're going to be the ones showing product teams how to use these traces anyways. 

> What would we need to do, besides pass a context?

Figure out an appropriate place to define a parent span, and then figure out how to pass the corresponding context around.

Reply:
i'm willing to bet something in this code base will have to change once we try to trace on the core side. in that light, i think it's fine make a best effort now and get the code merged

Reply:
> It feels like we're going in circles here, but it's probably my bad. I have no problem with including or not including this. It makes no difference to me. We're going to be the ones showing product teams how to use these traces anyways.

If it's the default suggestion and we know we need it later, then it seems like a no-brainer to include now.

> Figure out an appropriate place to define a parent span, and then figure out how to pass the corresponding context around.

I know that we can enhance things by adding our own ID at some outer level later on, but that's not what I'm worried about here. We already have calls like this, with context wired through:
```mermaid
sequenceDiagram
	core->>product: MethodA
	product->>relayer: MethodB
```
Complicated further by proxies:
```mermaid
sequenceDiagram
	core->>product: MethodA
	product->>core: MethodB(proxy)
	core->>relayer: MethodB(proxy)
```
I thought that whatever is auto-populated by GRPC would be wired through already.

Reply:
Resolved offline: parent sampler apparently not necessary for grpc client-server and nested span associations based on empirical evidence.

Reply:
I included it for now - based on our conversation offline I think it's possible we get more for free and in the worst case it won't sample less than we intend.